### PR TITLE
[5.6] Add memoize() helper function

### PR DESCRIPTION
**Pull request by calebporzio**
I often find myself using a simple memoize strategy like the following in my code (particularly around eloquent queries)

```
public function voteCount()
{
    static $cache;

    return $cache = $cache ?? $this->someExpensiveVoteCountQuery();
}
```

This PR introduces a helper function called `memoize()` to take care of this for us and cover some edge cases the aforementioned strategy doesn't handle well.

```
public function voteCount()
{
    return memoize(function() {
        return $this->someExpensiveVoteCountQuery();
    }
}
```

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Thanks for the PR. I don't personally think this belongs in the Laravel core I'm afraid.

**Comment by JosephSilber** [comment]
This looks a little similar to Spatie's [`once`](https://github.com/spatie/once) package, but with very different rules as to how to determine uniqueness.

**Comment by devcircus** [comment]
Taylor's [comment](https://twitter.com/taylorotwell/status/794670852285628416?s=21) about including a similar memoize in the framework. 

I use something similar. Don't know that it necessarily doesn't belong though.

**Comment by dwightwatson** [comment]
Taylor did make a PR to once that indicated he was considering pulling it in to the core (spatie/once#18). Would be nice to see that one go ahead.

**Comment by sisve** [comment]
Those calls to serialize() worries me. Wouldn't a safer approach be to introduce a `Cache::getOrCreate($key, $factoryCallback)` method?

**Comment by deleugpn** [comment]
@sisve that already exists. It's called `Cache::remember`. I guess the goal here is to have an `array` (memory) cache together with a default cache mechanism  (file, redis, etc). Maybe introducing a multi-driver cache would be better?

File: src/Illuminate/Support/helpers.php [hunk 1]
```diff
@@ -686,6 +686,32 @@ function last($array)
...
+
+        $reflector = new ReflectionFunction($callback);
+
+        $thumbprint = $reflector->getFilename()
+            .$reflector->getStartLine()
+            .$reflector->getEndLine()
+            .serialize($reflector->getParameters())
+            .serialize($reflector->getStaticVariables());
+
+        $hash = md5($thumbprint);
```
Comment (deleugpn) [inline src/Illuminate/Support/helpers.php]: `md5` has been voted to be entitely removed without deprecated warnings on https://wiki.php.net/rfc/release-md5-deprecation

**Review by deleugpn — CHANGES_REQUESTED** [review]

**Comment by sisve** [comment]
For testing purposes; how do I clear the cache of this memoize() helper?

File: src/Illuminate/Support/helpers.php (hunk 1 above)
Comment (m1guelpf) [reply src/Illuminate/Support/helpers.php]: @deleugpn The proposal of that RFC is to remove MD5 checksums for PHP downloads, not to remove the PHP `md5()` function.

Comment (deleugpn) [reply src/Illuminate/Support/helpers.php]: That's a shame 😞

**Comment by taylorotwell** [comment]
I like where your heart is at, but probably will just let people pull in spatie/once for now.

**Pull request closed without merging**